### PR TITLE
Add pylint

### LINKED ISSUE #8: Add pylint
Do we want to include code linters early on?
- Add pylint & pylint-django as dependency
- Decide on config, e.g. have a look at [integreat-cms](https://github.com/digitalfabrik/integreat-cms/blob/68a3810c3b20cc83d1fb93d71e6ed5af60f591dc/pyproject.toml#L249-L297)
- Add CircleCI job to test code against the linter

### PR DESCRIPTION
**Pull request by JoeyStk**
### Short description
This PR adds pylint to our project. I copied the settings we have for Integreat. I noticed that we have a lot more settings [disabled for Lunes](https://github.com/digitalfabrik/lunes-cms/blob/develop/.pylintrc) and therefore kept the more stricter settings we have in [Integreat](https://github.com/digitalfabrik/integreat-cms/blob/develop/pyproject.toml). Please feel free to comment

### Proposed changes

- Add settings to pyproject.toml
- Add Circle CI job
- Add tool to run pylint.sh

### Side effects

- I hope none

### Resolved issues

Fixes: #8

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -34,7 +34,10 @@ dynamic = ["version"]
 
 [project.optional-dependencies]
 dev = [
-    "build"
+    "build",
+    "pylint<=2.16.4",
```
Comment (timobrembeck) [inline pyproject.toml]:
```suggestion
    "pylint",
```
I think this constraint isn't required anymore? Did you copy and old state of the integreat config?

File: pyproject.toml [hunk 2]
```diff
@@ -34,7 +34,10 @@ dynamic = ["version"]
...
+    "pylint-django",
+    "pylint-runner",
```
Comment (timobrembeck) [inline pyproject.toml]:
```suggestion
```
Pylint runner isn't required anymore as well

File: pyproject.toml [hunk 3]
```diff
@@ -62,3 +65,53 @@ version = { attr = "integreat_compass.__version__" }
...
+disable = [
+    "consider-using-namedtuple-or-dataclass",
+    "consider-using-tuple",
+]
+ignore-paths = [
+  ".venv",
+  "node_modules",
+  "build",
+]
+extension-pkg-whitelist = ["lxml"]
```
Comment (timobrembeck) [inline pyproject.toml]:
```suggestion
```
As long as lxml isn't used, I don't see a reason for this

File: pyproject.toml [hunk 4]
```diff
@@ -62,3 +65,53 @@ version = { attr = "integreat_compass.__version__" }
...
+django-settings-module="integreat_cms.core.settings"
```
Comment (timobrembeck) [inline pyproject.toml]:
```suggestion
django-settings-module="integreat_compass.core.settings"
```

File: tools/pylint.sh [hunk 5]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
+
+# This script can be used to run the pylint_runner while ignoring migrations.
+
+# Import utility functions
+# shellcheck source=./tools/_functions.sh
+source "$(dirname "${BASH_SOURCE[0]}")/_functions.sh"
```
Comment (ztefanie) [inline tools/pylint.sh]: You need to fix this path to be `source "$(dirname "${BASH_SOURCE[0]}")/utils/_functions.sh"`

File: pyproject.toml [hunk 6]
```diff
@@ -62,3 +65,53 @@ version = { attr = "integreat_compass.__version__" }
...
+]
+disable = [
+    "consider-using-namedtuple-or-dataclass",
+    "consider-using-tuple",
+]
+ignore-paths = [
+  ".venv",
+  "node_modules",
+  "build",
+]
```
Comment (timobrembeck) [inline pyproject.toml:104]:
```suggestion
]
fail-on = "I"
```
fail on "useless-suppression"-warnings on the other hand I would enable.

File: pyproject.toml [hunk 7]
```diff
@@ -62,3 +65,53 @@ version = { attr = "integreat_compass.__version__" }
...
+    "missing-module-docstring",
+    "no-member",
+    "unsupported-binary-operation",
+]
+enable = [
+    "useless-suppression"
+]
+
+[tool.pylint.reports]
+output-format = "colorized"
```
Comment (timobrembeck) [inline pyproject.toml]:
```suggestion
output-format = "colorized"

```
I think new lines at the end of files is always a good practice

**Review by ztefanie — COMMENTED** [review]
Have you tried running the pylint.sh script? It is not working for me neither in the CI

File: tools/pylint.sh [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+require_installed
+
+# Run pylint
+echo "Starting code linting with pylint..." | print_info
+pylint .
+echo "✔ Linting finished" | print_success
```
Comment (timobrembeck) [inline tools/pylint.sh]:
```suggestion
echo "✔ Linting finished" | print_success

```

File: tools/pylint.sh (hunk 5 above)
Comment (timobrembeck) [inline tools/pylint.sh]:
```suggestion
# shellcheck source=./tools/utils/_functions.sh
source "$(dirname "${BASH_SOURCE[0]}")/utils/_functions.sh"
```

**Review by timobrembeck — CHANGES_REQUESTED** [review]

File: integreat_compass/__init__.py [hunk 9]
```diff
@@ -1 +1,4 @@
+"""
+Shows current version of Integreatcompass
+"""
 __version__ = "2023.3.0"
```
Comment (timobrembeck) [inline integreat_compass/__init__.py]: I think it's a bit misleading to reduce the docstring of the __init__ file to only the version, it is much more the entry point for the whole integreat compass package.

```suggestion
This package contains the Integreat Compass, a digital platform aimed at improving the integration of refugees and migrants into German society and the labor market 
"""

#: The current version of integreat-compass
__version__ = "2023.3.0"
```

File: tools/pylint.sh [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+
+# Import utility functions
+# shellcheck source=./tools/utils/_functions.sh
+source "$(dirname "${BASH_SOURCE[0]}")/utils/_functions.sh"
+
+require_installed
+
+# Run pylint
+echo "Starting code linting with pylint..." | print_info
+pylint ./integreat_compass/*.py
```
Comment (timobrembeck) [inline tools/pylint.sh]:
```suggestion
pylint .
```
I think using the `recursive` option makes more sense than path globbing

File: pyproject.toml [hunk 11]
```diff
@@ -62,3 +64,53 @@ version = { attr = "integreat_compass.__version__" }
...
+    "pylint.extensions.emptystring",
+    "pylint.extensions.for_any_all",
+]
+enable = [
+    "consider-using-augmented-assign",
+]
+disable = [
+    "consider-using-namedtuple-or-dataclass",
+    "consider-using-tuple",
+]
```
Comment (timobrembeck) [inline pyproject.toml:98]:
```suggestion
]
recursive = true
```

File: pyproject.toml [hunk 12]
```diff
@@ -34,7 +34,9 @@ dynamic = ["version"]
 
 [project.optional-dependencies]
 dev = [
-    "build"
+    "build",
+    "pylint",
+    "pylint-django",
```
Comment (timobrembeck) [inline pyproject.toml:39]: You need to update the pinned dependencies as well, by executing the `./tools/update_dependencies.sh` tool.

**Review by timobrembeck — CHANGES_REQUESTED** [review]

File: tools/pylint.sh (hunk 10 above)
Comment (JoeyStk) [reply tools/pylint.sh]: Yes, I just realized that myself :)

File: pyproject.toml (hunk 12 above)
Comment (JoeyStk) [reply pyproject.toml:39]: Oh, thank you for this hint!

File: tools/pylint.sh (hunk 5 above)
Comment (JoeyStk) [reply tools/pylint.sh]: Oh, yes, thank you!

**Review by timobrembeck — APPROVED** [review]
LGTM :+1:

File: tools/pylint.sh [hunk 13]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
+
+# This script can be used to run the pylint_runner while ignoring migrations.
```
Comment (timobrembeck) [inline tools/pylint.sh]:
```suggestion
# This script can be used to run pylint
```
Just a slight inaccuracy

Comment (JoeyStk) [reply tools/pylint.sh]: Fair enough

**Review by ztefanie — APPROVED** [review]
Nice work :rocket:

**Comment by JoeyStk** [comment]
Thank you, and thank you for all the support :pray:

**Pull request merged**